contracts/contracts/strategies/BaseConvexMetaStrategy.sol

```diff
@@ -0,0 +1,218 @@
+            .balanceOf(address(this));
+
+        if (metapoolGaugePTokens > 0) {
+            uint256 value = metapoolGaugePTokens.mulTruncate(
+                metapool.get_virtual_price()
+            );
+            balance += value;
+        }
+
+        uint256 assetDecimals = Helpers.getDecimals(_asset);
+        balance = balance.scaleBy(assetDecimals, 18) / THREEPOOL_ASSET_COUNT;
+    }
+
+    function _approveBase() internal override {
+        IERC20 pToken = IERC20(pTokenAddress);
+        // 3Pool for LP token (required for removing liquidity)
+        pToken.safeApprove(platformAddress, 0);
+        pToken.safeApprove(platformAddress, type(uint256).max);
+        // Gauge for LP token
+        metapoolLPToken.safeApprove(cvxDepositorAddress, 0);
+        metapoolLPToken.safeApprove(cvxDepositorAddress, type(uint256).max);
+        // Metapool for LP token
+        pToken.safeApprove(address(metapool), 0);
+        pToken.safeApprove(address(metapool), type(uint256).max);
+        // Metapool for Metapool main token
+        metapoolMainToken.safeApprove(address(metapool), 0);
+        metapoolMainToken.safeApprove(address(metapool), type(uint256).max);
+    }
+
+    /**
+     * @dev Get the index of the coin
+     */
+    function _getMetapoolCoinIndex(address _asset)
+        internal
+        view
+        returns (uint128)
+    {
+        for (uint128 i = 0; i < 2; i++) {
+            if (metapoolAssets[i] == _asset) return i;
+        }
+        revert("Invalid Metapool asset");
+    }
+
+    /**
+     * @dev Sets max withdrawal slippage that is considered when removing
+     * liquidity from Metapools.
+     * @param _maxWithdrawalSlippage Max withdrawal slippage denominated in
+     *        wad (number with 18 decimals): 1e18 == 100%, 1e16 == 1%
+     *
+     * IMPORTANT Minimum maxWithdrawalSlippage should actually be 0.1% (1e15)
+     * for production usage. Contract allows as low value as 0% for confirming
+     * correct behavior in test suite.
+     */
+    function setMaxWithdrawalSlippage(uint256 _maxWithdrawalSlippage)
+        external
+        onlyVaultOrGovernorOrStrategist
+    {
+        require(
+            _maxWithdrawalSlippage <= 1e17,
+            "Max withdrawal slippage needs to be between 0% - 10%"
+        );
+        emit MaxWithdrawalSlippageUpdated(
+            maxWithdrawalSlippage,
+            _maxWithdrawalSlippage
+        );
+        maxWithdrawalSlippage = _maxWithdrawalSlippage;
+    }
+
+    /**
+     * @dev Collect accumulated CRV and CVX and send to Harvester.
+     */
+    function collectRewardTokens()
+        external
+        override
+        onlyHarvester
+        nonReentrant
+    {
+        // Collect CRV and CVX
+        IRewardStaking(cvxRewardStakerAddress).getReward();
+        _collectRewardTokens();
+    }
+
+    /**
+     * @dev If x a negative number return 0 else return x
+     */
+    function toPositive(int256 x) internal pure returns (uint256) {
```

Comment:
I don't think this is a clear name, because there's many possible ways to convert a number to a positive number (The most obvious being to flip the sign.)

Reply:
good point, renamed to more conventional `max` function here: https://github.com/OriginProtocol/origin-dollar/pull/1002/commits/b826c7c58550e6b274bfe869e4a4a48f704623d0

Reply:
Looks good.